index/miner/onchain.go

```diff
@@ -0,0 +1,144 @@
+package miner
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/textileio/filecoin/lotus/types"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+)
+
+const (
+	fullRefreshThreshold = 100
+	heightOffset         = uint64(20)
+)
+
+// updateOnChainIndex updates on-chain index information when a new tipset
+func (mi *MinerIndex) updateOnChainIndex(new *types.TipSet) error {
+	log.Info("updating on-chain index...")
+	var chainIndex ChainIndex
+	newtsk := types.NewTipSetKey(new.Cids...)
+	ts, err := mi.store.LoadAndPrune(mi.ctx, newtsk, &chainIndex)
+	if err != nil {
+		return err
+	}
```

Comment:
Here we start leveraging the new `chainstore`.
Whenever a new Tipset is discovered from the chain, `updateOnChainIndex` is called with that as a parameter.

Calling `LoadAndPrune` is like saying: _Hey, give me the last saved state that is compatible with this new tipset and is on the same fork_. With that saved state, we know that we should re-apply changes from that point towards `new`.

Reply:
Kickass.